### ISSUE TITLE
SingleApplication::instanceStarted rarely triggers on Windows

### ISSUE THREAD

**Issue by nioncode** [itay-grudev/SingleApplication#121]
It seems like the `SingleApplication::instanceStarted` is only triggered very rarely on Windows (~20-30%). On Linux, everything works perfectly, but on Windows the signal is simply not emitted.

Sample code:
```cpp
#include <QApplication>
#include <QMainWindow>
#include <singleapplication.h>
#ifdef _WIN32
#include <Windows.h>
#endif

int main(int argc, char *argv[]) {
  SingleApplication app(argc, argv, true,
                        SingleApplication::Mode::User |
                            SingleApplication::Mode::SecondaryNotification);

  if (app.isSecondary()) {
#ifdef _WIN32
    AllowSetForegroundWindow(DWORD(app.primaryPid()));
    //app.sendMessage(QByteArray());
#endif
    return 0;
  }

  QMainWindow w;
  w.setFixedSize(300, 200);
  w.show();

  QObject::connect(&app, &SingleApplication::instanceStarted, &w, [&] {
    fprintf(stdout, "new instance started\n");
    fflush(stdout);
    w.show();
    w.activateWindow();
    w.raise();
  });

  return app.exec();
}
```

Interestingly, when I uncomment `app.sendMessage(QByteArray());`, then everything works fine and the signal is raised correctly in the primary instance.

Might be similar to #86, which was declared to be fixed since v3.1.5.

**Comment by itay-grudev** [comment]
I believe the new connection message send by the constructor and the custom message send afterwards clash in some way. This is likely related to #111, #68 and  #75 and #47 which now finally make sense to me.

My theory is that the two messages are flushed at the same time and received as a single message. Adding that sleep proposed in #75 and #47 now makes perfect sense as that would probably mitigate the issue. I will probably have add a structure to messages to identify their length and identify the following cases:


* Partial message received in a single call of readInitMessageBody()
* Multiple messages received in the same call of readInitMessageBody()

All of this is speculation. I currently don't have the time to investigate or fix the issue. I will not be able to review it the two weeks as I am otherwise occupied. If one of the other collaborators or you have the time - I would greatly appreciate it, alternatively - just give me some time. - I'm on it.

**Comment by nioncode** [comment]
Thanks for the fast feedback.

I did some rough tests and it seems like:

1. header (8 bytes) arrives first in one chunk and triggers `QLocalSocket::readyRead`
2. disconnect triggers `QLocalSocket::disconnected`, which removes the entry from the `connectionMap`
3. the message itself arrives as second chunk and triggers `QLocalSocket::readyRead` again -> here the info is fetched with `connectionMap[nextConnSocket]`, which then contains garbage data and does random stuff it (in my case this always results in calling `readInitMessageHeader` again, which then mistakenly takes the first 8 bytes from the message content and interprets it as header, which results in a `msgLen` of `1383560804`, which is obviously way too much data then what is available in the socket (which is `55`), thus the method does nothing further and waits for further data (which will never arrive) and does not raise the signal.

So in summary I think the only problem with the current code is that `disconnect` might be triggered before all data has been received and this messes up the partially stored information. A quick fix would be to simply wait for all data to be available as part of reading a single message, which would go against the event based approach, but might be neglectable compared to the complexity of implementing this in a event driven way.

**Comment by itay-grudev** [comment]
Interesting observation. I wouldn't even have thought that data could still be arriving after disconnect - it seems so counter intuitive. As per your solution - simply waiting for the whole message would end up blocking the main thread of your app and I think that shouldn't be acceptable.I will have to figure out a way to do the complicated solution.

**Comment by itay-grudev** [comment]
I think any data left should be read and processed in the `disconnected()` signal itself. If I understand the documentation correctly that should be enough.

**Comment by itay-grudev** [comment]
I still think that the partial message problem would still be there though.

**Comment by nioncode** [comment]
Maybe a good approach would be to send an acknowledgement message from the primary to the secondary instance and only close the socket in the secondary after the acknowledgement has been received.

The partial messages and `disconnect()` happening before `readyRead()` play together here, since in `disconnect()` the value of `sock->bytesAvailable()` is sometimes only 8 as well, so we cannot read all the remaining data here.
Thus I think it would be crucial to:
1. Support correctly receiving partial messages (a quick fix for this that seems to work is actually sending the header + initMsg not in two separate write calls, but generate a single `QByteArray` that contains both and send it in one `write` call; this might break for bigger messages though if the system decides to transmit them in chunks). This should also be supported for user defined messages, which currently do not have a header and are just pushed to the receiver however the system chunked the message and not in one single object.
2. Only close the socket after an acknowledgement has been received. This ensures that `disconnect()` is called only after everything has been received correctly. Maybe there is some easier way to achieve this without having to add an acknowledgement message, since it seems like this should be supported by sockets in general.

**Comment by itay-grudev** [comment]
I believe **1** is the better long-term solution albeit harder.

**Comment by nioncode** [comment]
Yes, but I think we need both, since `disconnect()` seems to sometimes fire before the whole message is received. I.e. you might have received 0 bytes (nothing), 8 bytes (header only), or all bytes and you cannot get the remaining ones in `disconnect()`. Maybe it works if we only delete the socket and remove the `ConnectionInfo` from the map when `disconnect()` has been triggered AND all bytes have been received.

**Comment by itay-grudev** [comment]
According to the documentation though, data may still be available in disconnect() and needs to be proceed. **What the library isn't currently doing - reading any available date in the slot.** It is safe to dispose of the socket afterwards.

**Comment by nioncode** [comment]
Maybe I did something wrong, but in my tests `bytesAvailable()` did not have the entire remaining data when called from `disconnect()`, so I would assume that you cannot read all remaining data simply from `disconnect()`, but need to wait until it has arrived. Maybe `bytesAvailable` is just not uptodate at this point in time though and `readAll` would actually return all data?

**Comment by itay-grudev** [comment]
I would just be speculating here and I am overwhelmed by work right now. I cannot continue this conversation without some proper investigation which I will do at a later time.

**Comment by dail8859** [comment]
Not sure if its related but I'm running into a situation where I am losing the first 8 bytes of the message being sent.

I'm always getting the expected signals but just missing some data. 

If this seems unrelated I can open a new issue.

**Comment by nioncode** [comment]
After reviewing #125 (and when this gets merged), I think the issue here boils down to the sender closing the socket too early (as part of the destructor) when not all data has been sent, yet.

I think the problem comes from the fact that `waitForBytesWritten` and `flush` actually might not send ALL data that is currently 'waiting' on the socket, but only some. E.g. when calling those in a loop until `bytesToWrite() == 0`, then it works fine in my test case. This might cause blocking, so I don't know if this is acceptable as part of the send methods, but we could at least modify the destructor to make sure all buffered data has been sent:

```cpp
SingleApplicationPrivate::~SingleApplicationPrivate()
{
    if( socket != nullptr ){
        while (socket->bytesToWrite() != 0) {
            socket->waitForBytesWritten(5);
            socket->flush();
        }
        socket->close();
        delete socket;
    }
...
```

With this patch, I did not see any missing `instanceStarted` signals.

**Comment by Shatur95** [comment]
Hmmm, this is weird. The current solution doesn't look very good. Is there some way to delay the destruction of the `SingleApplicationPrivate`?

**Comment by dail8859** [comment]
From the [documentation](https://doc.qt.io/qt-5/qlocalsocket.html#disconnectFromServer) it looks like `QLocalSocket::disconnectFromServer()` is a more graceful way to end a connection...and then just wait for the signal.

> Attempts to close the socket. If there is pending data waiting to be written, QLocalSocket will enter ClosingState and wait until all data has been written. Eventually, it will enter UnconnectedState and emit the disconnectedFromServer() signal.

**Comment by nioncode** [comment]
This was fixed by #133.
